### DISCUSSION TITLE
[Feature] more flexible routing / catch-all route with SSG and SSR

### DISCUSSION THREAD

**Discussion by Zauberfisch** [vercel/next.js#17987] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

All my routes are coming from an API/CMS. The Homepage, thousands of content pages, Registration Page, Login-Only Pages, and so on ...
An example of what that data in the database might look like:
```json
[
  { "lang": "en", "path": "/", "type": "HomePage", "title": "Home", "navTitle": "Home", "content": "<p>....." },
  { "lang": "en", "path": "/about-us", "type": "DefaultPage", "title": "About this webiste", "navTitle": "About", "content": "<p>....." },
  { "lang": "en", "path": "/create-account", "type": "RegisterPage", "title": "Create an Account", "navTitle": "Signup", "content": "<p>.....", "contentAfterFormSubmission": "Thanks! Check your mail"  },
  { "lang": "en", "path": "/secret-page", "type": "DefaultPage", "title": "Secret information", "navTitle": "Secret", "content": "<p>.....", "requiresLogin": true },
  { "lang": "en", "path": "/some-page-that-does-serverside-stuff", "type": "ServerSideCalculationPage", "title": "Showing Realtime Data", "navTitle": "Realtime Data", "content": "<p>....." },

  { "lang": "de", "path": "/de", "type": "HomePage", "title": "Home", "navTitle": "Home", "content": "<p>....." },
  { "lang": "de", "path": "/de/ueber-uns", "type": "DefaultPage", "title": "Ueber diese Website", "navTitle": "Ueber uns", "content": "<p>....."},
  { "lang": "de", "path": "/de/account-erstellen", "type": "RegisterPage", "title": "Account erstellen", "navTitle": "Signup", "content": "<p>.....", "contentAfterFormSubmission": "Danke! ..." },
  { "lang": "de", "path": "/de/geheime-seite", "type": "DefaultPage", "title": "Geheime Information", "navTitle": "Geheim", "content": "<p>.....", "requiresLogin": true },
]
```
As this example shows, the content authors have full control over the path structure, amount of pages and used languages.  
So even special pages like a registration page can not be a static path inside `pages/`.  

## Describe the solution you'd like

I can think of 3 ways:

1. **getRoutes in next.config.js**
   add a callback `getRoutes` to `next.config.js` (or other place) which returns a map of routes with the component/template to use. Maybe something like this:
   ```js
   // next.config.js
   getRoutes: async () => {
      // some API call here
     return {
       routes: [
            // 3 routes with getServerSideProps
            { params: {slug: ['secret-page']}, template: `components/my-pages/DefaultPage.js`, getServerSideProps: () => { /* do something here */ } },
            { params: {slug: ['some-page-that-does-serverside-stuff']}, template: `components/my-pages/SpeicalSSRPage.js`, getServerSideProps: () => { /* do something here */ } },
            { params: {slug: ['de', 'geheime-seite']}, template: `components/my-pages/DefaultPage.js`, getServerSideProps: () => { /* do something here */ } },

            // 6 routes with getStaticProps
            { params: {slug: []}, template: `components/my-pages/HomePage.js`, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['about-us']}, template: `components/my-pages/DefaultPage.js`, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['create-account']}, template: `components/my-pages/CreateAccount.js`, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de']}, template: `components/my-pages/HomePage.js`, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de', 'ueber-uns']}, template: `components/my-pages/DefaultPage.js`, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de', 'account-erstellen']}, template: `components/my-pages/CreateAccount.js`, getStaticProps: () => { return { /* some data */ } } },
       ], 
       fallback: false
     };
   }
   ```
    
   This way I could just set `useFileSystemPublicRoutes: false` to disable `pages` and only define my routes here.  
   Ideally we'd also have some mechanism to specify how often this is reevaluated to add new pages that might have been added in the CMS or API.   

1. **getRoutes in a page**
   Basically the same thing as above, but instead of inside `next.config.js` inside a page:
   replace `getStaticPaths` with `getRoutes` that can return routes with either `getStaticProps` or `getServerSideProps` callbacks like in Option 1 above.
   ```js
   // pages/[[...slug]].js
   export default function Page() { /* ... template to be used has to be decided based on props ... */ }
   export async function getRoutes() {
      // some API call here
     return {
       routes: [
            // 3 routes with getServerSideProps
            { params: {slug: ['secret-page']}, getServerSideProps: () => { /* do something here */ } },
            { params: {slug: ['some-page-that-does-serverside-stuff']}, getServerSideProps: () => { /* do something here */ } },
            { params: {slug: ['de', 'geheime-seite']}, getServerSideProps: () => { /* do something here */ } },

            // 6 routes with getStaticProps
            { params: {slug: []}, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['about-us']}, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['create-account']}, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de']}, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de', 'ueber-uns']}, getStaticProps: () => { return { /* some data */ } } },
            { params: {slug: ['de', 'account-erstellen']}, getStaticProps: () => { return { /* some data */ } } },
       ], 
       fallback: false
     };
   }
   ```

1. **getServerSidePaths**
   add `getServerSidePaths` and allow both `getStaticPaths`+`getStaticProps` and `getServerSideProps`+`getServerSidePaths` in one file
    The way I imagine this:
    1. Step 1: call `getServerSidePaths`, if route matches a path here, use `getServerSideProps`
    1. Step 2: if not, then call `getStaticPaths`, if route matches a path here, use `getStaticProps`
    1. Step 3: if no patch matches, then consider the value of `fallback` from `getStaticPaths` and proceed accordingly
   ```js
   // pages/[[...slug]].js
   export default function Page() { /* ... template to be used has to be decided based on props ... */ }
   export async function getServerSideProps() { /* ... */ }
   export async function getStaticProps() { /* ... */ }
   export async function getServerSidePaths() { 
      return { 
         paths: [
            { params: {slug: ['secret-page']} },
            { params: {slug: ['some-page-that-does-serverside-stuff']} },
            { params: {slug: ['de', 'geheime-seite']} },
         ]
      };
   }
   export async function getStaticPaths() {
      return {
         paths: [
            { params: {slug: []} },
            { params: {slug: ['about-us']} },
            { params: {slug: ['create-account']} },
            { params: {slug: ['de']} },
            { params: {slug: ['de', 'ueber-uns']} },
            { params: {slug: ['de', 'account-erstellen']} },
         ],
         fallback: false
      };
   }
   ```
   This isn't fully thought through yet though, the way I propose SSR routes go first, then static, then fallback. But a potential usecase could also be that someone wants SSG first and SSR as fallback.

## Describe alternatives you've considered

1. **Only SSG (with extra api and client side rendering for private routes)**:
   Just a single file in pages: `pages/[[...slug]].js` with `getStaticPaths`+`getStaticProps` for all public routes and then do the login-only routes via an extra client facing API and client side only rendering.

   Major downside: SSR not possible at all

1. **Only SSR**:
   Just a single file in pages: `pages/[[...slug]].js` with `getServerSideProps` for all routes (public and login-only).

   Major downside: (built-in) SSG not possible at all

1. **Only SSR with an extra cache server that serves some pages static**:
   Possible, but next has builtin SSG functionality, and it would be nice if I can use it

   Major downside: requires an additional server/service and configuration to enable a functionality that already exists in next

1. **Prefix SSR or SSG paths**:
   I could work around this problem by creating `pages/[[...ssgSlug]].js` and `pages/ssr/[[...ssrSlug]].js` and change the API to serve all pages that need ssr with a prefix `ssr/` in the path. 

   Major downside: less friendly/weird URLs

1. **Custom server with `useFileSystemPublicRoutes: false`**:
   This would enable me to define route the way I want.  
   I am not yet a 100% sure what side effects this will bring, but from what I understand I will loose some optimisations that usually come builtin with nextjs.

1. **Only CSR**:
   Just adding this so nobody suggests it as a solution. Obviously not desirable.

**Comment by Zauberfisch** [comment]
on second thought, custom server might be what I am after. I was just discouraged from looking into it by the warnings that I will loose access to features.  
I guess using caching in the custom server (like the SSR Caching example) or a cache in front of nextjs will result in a very similar end result as SSG for pages that get cached.

But I think I still prefer the "getRoutes in next.config.js" approach I suggested in my initial post.

**Reply by Zauberfisch** [reply]
I've now taken an in-depth look at custom server, but I've not been able to achieve my desired behaviour. I've created a bug report related to that here: #18070

**Comment by Zauberfisch** [comment]
I've just come across a related RFC: https://github.com/vercel/next.js/discussions/9081 and https://nextjs.org/docs/api-reference/next.config.js/rewrites which I previously did not find.

Unfortunately, it is still limited in it's scope and will not work for this use case: as far as I understand rewrites are defined at build time and therefore pages created in the CMS after the time of build will not be routeable, so I will keep this discussion/issue open.

**Comment by macdermott** [comment]
Hey @Zauberfisch. Did you ever find a workable solution for this problem?

**Reply by Zauberfisch** [reply]
@macdermott no. 
For a while I used just a single page (`pages/[[...slug]].js`) and then implemented my own custom routing and template finding logic.
But eventually came to the conclusion that nextjs is not ready for more complex websites. I've spent so much time working around weird limitations and bugs* that I eventually just decided it's not worth it for SSR. Now I am just using a well optimised CSR setup.


\* (custom server was basically unusable at the time I tried it, trailing-slash redirects were buggy but you couldn't turn them off, ...)
